### DISCUSSION TITLE
Case-Insensitive Languages: Any Tips?

### DISCUSSION THREAD

**Discussion by tajmone** [tree-sitter/tree-sitter#1661] [Q&A]
I've read Discussions #1063 and #1342, which touch on the topic of handling case-insensitive language/syntaxes grammars, and saw the solution adopted by the [Fortran grammar] via its custom [`caseInsensitive()`][Fortran grammar.js] function which leverages Tree-sitter's native `alias(rule, string literal)`.

But since the official documentation doesn't mention the topic of case-insensitive languages (although there are a many such languages) I was wondering if anyone could supply some advise and tips on the best way to handle languages where all tokens are case-insensitive, without affecting the performance of the generated parser.

Wouldn't it be worth to add some native support for case-insensitive keywords?

If I've understood correctly some of the responses in the mentioned Discussions, the general advise for non-keywords is to match them as identifiers, and then handle them individually later (although it wasn't clear to me yet how that would need to be done) — [quoting from #1063]:

> @maxbrunsfeld — Ok, so distinguishing these should not be the parser's job. Just parse them all as `identifier` (or something) in your grammar. Then, when analyzing your code, check the text of the `identifier` tokens, and handle built-in identifiers based on their text.

I'm not sure what "analyzing your code" means here — within the `grammar.js` file? via queries? in an external scanner? I've just started with Tree-sitter, so I'm still trying to the get the hang of it.

I'm trying to write a custom grammar for a language that is _entirely_ case insensitive, and where users are likely to use all sort of letter-case variations since it has a prose-oriented syntax that encourages users to code as if they were writing natural sentences.

So far, I noticed that the `i` RegEx modifier (e.g. `/some-regex/i`) seems to be ignored by Tree-sitter, for it doesn't match case-insensitively.

Any suggestions?

[Fortran grammar]: https://github.com/stadelmanma/tree-sitter-fortran

[Fortran grammar.js]: https://github.com/stadelmanma/tree-sitter-fortran/blob/a9c79b2/grammar.js#L1125

[quoting from #1063]: https://github.com/tree-sitter/tree-sitter/discussions/1063#discussioncomment-628269

**Comment by dfgordon** [comment]
The definition of the term "parser" is the key to understanding what is said there.  I think he intends a strict definition where the parser is only that part of Tree-sitter that transforms a source file into a syntax tree.

If this is right, then "analyzing your code" is anything you do with the syntax tree once you've got it.   For example, walk the tree using a cursor, and when you hit a node of type `identifier` grab the text of the node (the text that is actually in the source file) to make further distinctions about what it is.

Queries are a way to search the tree without having to walk it manually, so that could be an option depending on what you want to do.

Both `grammar.js` and external scanners are deeply involved with producing the syntax tree, so the quote recommends you do *not* use these to break out different meanings of identifiers.
